File: leetcode/LRUCache.java
import java.util.HashMap;
class LRUCache {

    private Node head;
    private Node tail;
    private int capacity; 
    private HashMap<Integer,Node> c;
    
    public LRUCache(int capacity) {
        tail = new Node();
        head = new Node();
        head.next = tail;
        tail.prev = head;
        this.capacity = capacity; 
        this.c = new HashMap<Integer, Node>(capacity);
    }
    
    public int get(int key) {
        

        Node node = this.c.get(key);
        
        if(node == null){
            return -1;
        };
                        System.out.print(node.prev);
                        System.out.print(node.next);

        remove(node);
        add(node);
        return node.value; 
    }
    
    public void put(int key, int value) {
        if(c.containsKey(key)){
            Node node = c.get(key);
            node.value = value;
            remove(node);
            add(node);
        }
        else {
            if(c.size() == capacity){
                int tailkey = tail.prev.key;
                c.remove(tailkey);
                remove(tail.prev);
            }
            Node newNode = new Node(key, value);
            c.put(key, newNode);
            add(newNode);
        }
    }
    private void add(Node node){
        Node nextHead = head.next;
        nextHead.prev = node;
        node.next = nextHead; 
        node.prev = head;
        head.next = node; 
    }
    private  void remove(Node node){
        Node prev = node.prev;
        Node next = node.next;
        next.prev = prev;
        prev.next = next;
    }
    
    class Node {
        public int key;
        public int value; 
        public Node next;
        public Node prev;
        public Node(){
            
        }
        public Node(Integer key, Integer value){
            this.key = key;
            this.value = value;
        }
    }
    
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: leetcode/LRUcache.js
/**
 * @param {number} capacity
 */

class Node {
  constructor(key, value) {
    this.key = key;
    this.value = value;
    this.next = null;
    this.prev = null;
  }
}
var LRUCache = function (capacity) {
  this.capacity = capacity;
  this.hash = {};
  this.head = new Node();
  this.tail = new Node();
  this.head.next = this.tail;
  this.tail.prev = this.head;
  this.currentTenancy = 0;
};

/**
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function (key) {
  if (!this.hash.hasOwnProperty(key)) {
    return -1;
  }
  let node = this.hash[key];
  this.remove(node);
  this.add(node);
  return node.value;
};
LRUCache.prototype.add = function (node) {
  let headNext = this.head.next;
  this.head.next = node;
  node.prev = this.head;
  node.next = headNext;
  headNext.prev = node;
};
LRUCache.prototype.remove = function (node) {
  let prev = node.prev;
  let next = node.next;
  prev.next = next;
  next.prev = prev;
};

/**
 * @param {number} key
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function (key, value) {
  if (this.hash.hasOwnProperty(key)) {
    let node = this.hash[key];
    node.value = value;
    this.remove(node);
    this.add(node);
  } else {
    if (this.currentTenancy == this.capacity) {
      let lastNode = this.tail.prev;
      delete this.hash[lastNode.key];
      this.remove(lastNode);
      this.currentTenancy--;
    }
    let newNode = new Node(key, value);
    this.hash[key] = newNode;
    this.add(newNode);
    this.currentTenancy++;
    console.log(this.currentTenancy);
  }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
